// @roots/bud-dashboard/lib/cjs/Dashboard/line.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Line = void 0;
const bud_support_1 = require("@roots/bud-support");
const spinner_1 = require("./spinner");
class Line {
    constructor(spinner = true) {
        this.isComplete = false;
        if (spinner)
            this.spinner = new spinner_1.Spinner();
    }
    get icon() {
        if (!this.spinner || this.isComplete)
            return null;
        return this.spinner.frame;
    }
    get frame() {
        return this.icon ? [this.icon, this.text].join(' ') : this.text;
    }
    update(...text) {
        this.text = text[0] ? text?.join(' ') : null;
        return this;
    }
    complete(isComplete = true) {
        this.isComplete = isComplete;
        return this;
    }
}
__decorate([
    bud_support_1.bind
], Line.prototype, "update", null);
__decorate([
    bud_support_1.bind
], Line.prototype, "complete", null);
exports.Line = Line;
