// @roots/bud-framework/lib/cjs/extension/decorators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.when = exports.value = exports.production = exports.plugin = exports.options = exports.label = exports.expose = exports.development = exports.dependsOn = exports.dependsOnOptional = exports.bind = void 0;
var bud_support_1 = require("@roots/bud-support");
Object.defineProperty(exports, "bind", { enumerable: true, get: function () { return bud_support_1.bind; } });
var dependsOnOptional_1 = require("./dependsOnOptional");
Object.defineProperty(exports, "dependsOnOptional", { enumerable: true, get: function () { return dependsOnOptional_1.dependsOnOptional; } });
var dependsOn_1 = require("./dependsOn");
Object.defineProperty(exports, "dependsOn", { enumerable: true, get: function () { return dependsOn_1.dependsOn; } });
var development_1 = require("./development");
Object.defineProperty(exports, "development", { enumerable: true, get: function () { return development_1.development; } });
var expose_1 = require("./expose");
Object.defineProperty(exports, "expose", { enumerable: true, get: function () { return expose_1.expose; } });
var label_1 = require("./label");
Object.defineProperty(exports, "label", { enumerable: true, get: function () { return label_1.label; } });
var options_1 = require("./options");
Object.defineProperty(exports, "options", { enumerable: true, get: function () { return options_1.options; } });
var plugin_1 = require("./plugin");
Object.defineProperty(exports, "plugin", { enumerable: true, get: function () { return plugin_1.plugin; } });
var production_1 = require("./production");
Object.defineProperty(exports, "production", { enumerable: true, get: function () { return production_1.production; } });
var value_1 = require("./value");
Object.defineProperty(exports, "value", { enumerable: true, get: function () { return value_1.value; } });
var when_1 = require("./when");
Object.defineProperty(exports, "when", { enumerable: true, get: function () { return when_1.when; } });
